router/src/server.rs

```diff
@@ -494,7 +515,146 @@ async fn generate_stream(
+    Extension(infer): Extension<Infer>,
+    Json(req): Json<ChatRequest>,
+) -> Result<Response, (StatusCode, Json<ErrorResponse>)> {
+    metrics::increment_counter!("tgi_request_count");
+
+    // extract the values we need for the chat request
+    let stream = req.stream;
+    let max_new_tokens = match req.max_tokens {
+        Some(max_new_tokens) => Some(max_new_tokens),
+        None => Some(100),
+    };
+
+    // apply chat template to flatten the request into a single input
+    let inputs = match infer.apply_chat_template(req) {
+        Ok(inputs) => inputs,
+        Err(err) => {
+            metrics::increment_counter!("tgi_request_failure", "err" => "validation");
+            tracing::error!("{err}");
+            return Err((
+                StatusCode::UNPROCESSABLE_ENTITY,
+                Json(ErrorResponse {
+                    error: err.to_string(),
+                    error_type: err.error_type().to_string(),
+                }),
+            ));
+        }
+    };
+
+    // poor man's token count (assumes that each character is a token)
+    let prompt_character_count: u32 = inputs.chars().count().try_into().unwrap_or_default();
+
+    // build the request passing some parameters
+    let generate_request = GenerateRequest {
+        inputs: inputs.to_string(),
+        parameters: GenerateParameters {
+            best_of: None,
+            temperature: None,
+            repetition_penalty: None,
+            top_k: None,
+            top_p: None,
+            typical_p: None,
+            do_sample: false,
+            max_new_tokens,
+            return_full_text: None,
+            stop: Vec::new(),
+            truncate: None,
+            watermark: false,
+            details: true,
+            decoder_input_details: false,
+            seed: None,
+            top_n_tokens: None,
+        },
+    };
+
+    // switch on stream
+    if stream {
+        let stream_count = AtomicU32::new(0);
+        // pass this callback to the stream generation and build the required event structure
+        let on_message_callback = move |stream_token: StreamResponse| {
+            let event = Event::default();
+
+            let current_time = std::time::SystemTime::now()
+                .duration_since(std::time::UNIX_EPOCH)
+                .unwrap_or_else(|_| std::time::Duration::from_secs(0))
+                .as_secs();
+
+            // increment the stream count
+            stream_count.fetch_add(1, Ordering::SeqCst);
+            let current_stream_count = stream_count.load(Ordering::SeqCst);
+
+            event
+                .json_data(ChatCompletionChunk::new(
+                    stream_token.token.text,
+                    current_time,
+                    current_stream_count,
+                ))
+                .unwrap_or_else(|_| {
+                    println!("Failed to serialize ChatCompletionChunk");
+                    Event::default()
+                })
+        };
```

Comment:
Maybe this can be refactored so that the idx of the stream_token is part of the signature of the callback instead of relying on an atomic? It can also be part of the StreamResponse struct.